### ISSUE TITLE
您好我在跑这个etth的插补这一块的代码，可视化部分是按照这个吗

### ISSUE THREAD

**Issue by lyli17** [Y-debug-sys/Diffusion-TS#87]
plt.rcParams["font.size"] = 12
fig, axes = plt.subplots(nrows=5, ncols=1, figsize=(12, 15))

ori_data = np.load(os.path.join(dataset.dir, f"sine_ground_truth_{seq_length}_test.npy"))
# ori_data = np.load(os.path.join(dataset.dir, f"{dataset_name}_norm_truth_{seq_length}_test.npy"))  # Uncomment the line if dataset other than Sine is used.
masks = np.load(os.path.join(dataset.dir, f"sine_masking_{seq_length}.npy"))
sample_num, seq_len, feat_dim = ori_data.shape
observed = ori_data * masks

for feat_idx in range(feat_dim):
    df_x = pd.DataFrame({"x": np.arange(0, seq_len), "val": ori_data[0, :, feat_idx],
                         "y": masks[0, :, feat_idx]})
    df_x = df_x[df_x.y!=0]

    df_o = pd.DataFrame({"x": np.arange(0, seq_len), "val": ori_data[0, :, feat_idx],
                         "y": (1 - masks)[0, :, feat_idx]})
    df_o = df_o[df_o.y!=0]
    axes[feat_idx].plot(df_o.x, df_o.val, color='b', marker='o', linestyle='None')
    axes[feat_idx].plot(df_x.x, df_x.val, color='r', marker='x', linestyle='None')
    axes[feat_idx].plot(range(0, seq_len), samples[0, :, feat_idx], color='g', linestyle='solid', label='Diffusion-TS')
    plt.setp(axes[feat_idx], ylabel='value')
    if feat_idx == feat_dim-1:
        plt.setp(axes[-1], xlabel='time')

plt.show()

**Comment by Y-debug-sys** [comment]
你好，论文中的可视化主要参考 https://github.com/ermongroup/CSDI/blob/main/visualize_examples.ipynb 。

**Comment by lyli17** [comment]
好的谢谢您提供的代码，还想问一下就是我复现这个imputation的时候，我查看得到的这个npy文件发现masks和gen_data的维度与ori_data维度不一致，但是参考了那个Tutorial_1里的代码observed=ori_data * masks这两个维度不一致相乘会报错，不知道是不是我这里运行出错还是什么原因，辛苦解答一下，谢谢~
![图片1](https://github.com/user-attachments/assets/03f103d0-b354-4830-8ced-e734a61419db)
![图片2](https://github.com/user-attachments/assets/b046e08c-e4fe-4432-9606-8864df273e26)

**Comment by Y-debug-sys** [comment]
请问这三个变量分别是在什么地方产生的？

**Comment by lyli17** [comment]
是这个：
ori_data = np.load(os.path.join('D:\\Diffusion-TS\\Diffusion-TS-main\\OUTPUT\\11\\samples', f'etth_norm_truth_24_test.npy'))

gen_data = np.load(os.path.join('D:\\Diffusion-TS\\Diffusion-TS-main\\OUTPUT\\11', f'ddpm_infill_gjl11.npy'))

masks = np.load(os.path.join('D:\\Diffusion-TS\\Diffusion-TS-main\\OUTPUT\\11\\samples', f'etth_masking_24.npy'))

**Comment by lyli17** [comment]
这三个变量是这个：
ori_data = np.load(os.path.join('D:\\gitcode\\Diffusion-TS\\Diffusion-TS-main\\OUTPUT\\11\\samples', f'etth_norm_truth_24_train.npy'))

gen_data = np.load(os.path.join('D:\\Diffusion-TS\\Diffusion-TS-main\\OUTPUT\\11', f'ddpm_infill_11.npy'))

masks = np.load(os.path.join('D:\\Diffusion-TS\\Diffusion-TS-main\\OUTPUT\\11\\samples', f'etth_masking_24.npy'))

**Comment by lyli17** [comment]
这三个变量是这个：
ori_data = np.load(os.path.join('D:\Diffusion-TS\Diffusion-TS-main\OUTPUT\11\samples', f'etth_norm_truth_24_train.npy'))

gen_data = np.load(os.path.join('D:\Diffusion-TS\Diffusion-TS-main\OUTPUT\11', f'ddpm_infill_11.npy'))

masks = np.load(os.path.join('D:\Diffusion-TS\Diffusion-TS-main\OUTPUT\11\samples', f'etth_masking_24.npy'))

**Comment by lyli17** [comment]
不好意思打错啦，就是test之后还是对不上

**Comment by Y-debug-sys** [comment]
我也不太清楚，按照代码逻辑，正常运行 test 指令，ori_data 和 masks 的 shape 一定是一样的。

**Comment by lyli17** [comment]
我是这么运行的：1.python main.py --name 1.7test --config_file Config/etth.yaml --gpu 0 --train
2.python main.py --name 1.7test --config_file Config/etth.yaml --gpu 0 --sample 1 --milestone 5 --mode infill --missing_ratio 0.1                                   你看看是哪里操作的不对吗，我也不清楚为什么这个shape不一致，然后etth.yaml里我把train和test比例改了7：3；
dataloader:
  train_dataset:
    target: Utils.Data_utils.real_datasets.CustomDataset
    params:
      name: etth
      proportion: 0.7  # Set to rate < 1 if training conditional generation
      data_root: ./Data/datasets/ETTh.csv
      window: 48  # seq_length
      save2npy: True
      neg_one_to_one: True
      seed: 123
      period: train

  test_dataset:
    target: Utils.Data_utils.real_datasets.CustomDataset
    params:
      name: etth
      proportion: 0.3  # rate
      data_root: ./Data/datasets/ETTh.csv
      window: 48  # seq_length
      save2npy: True
      neg_one_to_one: True
      seed: 123
      period: test
      style: separate
      distribution: geometric
    coefficient: 1.0e-2
    step_size: 5.0e-2
    sampling_steps: 200

  batch_size: 128
  sample_size: 256
  shuffle: False

**Comment by Y-debug-sys** [comment]
这两个 proportion 必须一致。

**Comment by lyli17** [comment]
> 这两个 proportion 必须一致。

好的谢谢，还有一个疑问，就是关于这个Standard TS Generation和conditional TS generation这两个任务有什么区别呢，我有点没读懂，后者就是既是包含imputation任务也可以表示prediction任务对吗，前者就是无条件的prediction任务吗，感觉理解有误，希望可以帮忙解答一下疑问。另一个就是imputation任务除了对比csdi模型还有其他什么baseline可以对比的嘛

**Comment by Y-debug-sys** [comment]
> 好的谢谢，还有一个疑问，就是关于这个Standard TS Generation和conditional TS generation这两个任务有什么区别呢，我有点没读懂，后者就是既是包含imputation任务也可以表示prediction任务对吗，前者就是无条件的prediction任务吗，感觉理解有误，希望可以帮忙解答一下疑问。另一个就是imputation任务除了对比csdi模型还有其他什么baseline可以对比的嘛

区别仅仅在于推测时有无条件；预测和插入本质都是条件生成（条件为观察值）；baseline 详见论文，更多的可以检索相关机器学习会议。

**Comment by lyli17** [comment]
好的，那前面的那个无条件生成就是standard ts generation 和long-term Ts generation 属于什么任务呀，相当于长时预测和短时预测嘛，我现在就是有点混淆这个生成任务和插补任务的区别

---- 回复的原邮件 ----
| 发件人 | Xinyu ***@***.***> |
| 发送日期 | 2025年01月07日 22:10 |
| 收件人 | Y-debug-sys/Diffusion-TS ***@***.***> |
| 抄送人 | lyli17 ***@***.***>,
Author ***@***.***> |
| 主题 | Re: [Y-debug-sys/Diffusion-TS] 您好我在跑这个etth的插补这一块的代码，可视化部分是按照这个吗 (Issue #87) |

好的谢谢，还有一个疑问，就是关于这个Standard TS Generation和conditional TS generation这两个任务有什么区别呢，我有点没读懂，后者就是既是包含imputation任务也可以表示prediction任务对吗，前者就是无条件的prediction任务吗，感觉理解有误，希望可以帮忙解答一下疑问。另一个就是imputation任务除了对比csdi模型还有其他什么baseline可以对比的嘛

区别仅仅在于推测时有无条件；预测和插入本质都是条件生成（条件为观察值）；baseline 详见论文，更多的可以检索相关机器学习会议。

**Comment by starry-ZRY** [comment]
想问一下作者大大，Figure 16用的sample_idx是什么

**Comment by Y-debug-sys** [comment]
> 想问一下作者大大，Figure 16用的sample_idx是什么

代码根据 https://github.com/ermongroup/CSDI/blob/main/visualize_examples.ipynb 改动而来。
